feat: impl ng `flox publish`

ysndr opened the pull request:
This PR implements the `publish2` subcommand using the `Publish` abstraction.

*Usage for reference*

```
build and publish project to flox channel

Usage: [-k ARG] [-c ARG] [-s ARG] [--prefer-https] [<INSTALLABLE>] <args>...

Available options:
    -k, --sign-key <ARG>         Signing key file to sign the binary with
                                 When omitted, reads from the config
    -c, --cache-url <ARG>        Url of a binary cache to push binaries _to_
                                 When omitted, reads from the config
    -s, --substituter-url <ARG>  Url of a substituter to pull binaries _from_
                                 When omitted, uses the value for cache-url and falls back to 
                                 the config
        --prefer-https           Prefer https access to repositories published with a `github:` 
                                 reference
    -h, --help                   Prints help information
```

ysndr commented:
This should be ready for review now

mkenigs commented on crates/flox/src/commands/package.rs:
```diff
@@ -205,6 +208,34 @@ pub(crate) mod interface {
+        ///
+        /// When ommitted, uses the value for cache-url and falls back to the config
+        #[bpaf(long, short('s'))]
+        pub substituter_url: Option<SubstituterUrl>,
```
Optional (bikeshed warning): any reason to rename these from `--upload-to` and `--download-from`? Those feel like words anyone will understand, which is not true of substituter.

ysndr replied:
Apart from that I find the the names a bit clunky, the similarity of the names suggests that I need to set both, at which point you still have to explain the differences.
`upload_to` and `download_from` also make weird config keys (verbs, rather than nouns) that require additional context.
For them to work in the config it would need to be `publish.upload_to` I think.

I used "cache" and "substituter" because those seem more descriptive depending on your position; As a developer you would want to cache binaries, while at as a consumer you'd want them to be provided by a substituter.
If we don't want to use the word substituter for some reason, it might be something like `providing-cache-url`?

But to cut the bikeshedding short, either way seems somewhat reasonable, I don't have strong preferences except against putting `upload_to` into the config without context

mkenigs commented on crates/flox/src/commands/package.rs:
```diff
@@ -350,22 +381,100 @@ impl PackageCommands {
+                    println!("{}", serde_json::to_string(analysis)?);
+                } else {
+                    info!("Uploading snapshot to {}...", publish_flakeref.clone_url());
+                    publish.push_snapshot().await.context("Failed to upload")?;
```
```suggestion
                    publish.push_snapshot().await.context("Failed to upload")?;
                    info!("done!");
```

mkenigs commented on crates/flox-rust-sdk/src/models/publish.rs:
```diff
@@ -226,22 +231,29 @@ impl<'flox> Publish<'flox, NixAnalysis> {
+        let command = Build {
+            installables: [self.installable()].into(),
+            ..Default::default()
+        };
```
Do we need to add the same `--override-input`s as in `get_drv_metadata`?

mkenigs replied:
Same question for `sign_binary()`. I wonder if it would be less error prone to use store_paths

ysndr replied:
Yeah here we need that, because we're not in the `crates/flox-rust-sdk/src/actions/package.rs` module which had a helper for that iirc.

sign binary and upload binary would only need that if we do use the `^*` installable (since that will cause an evaluation, unlike store paths)
Maybe that's a reason _against_ using `^*` there?

ysndr replied:
generally we should set that default override earlier 🤔

mkenigs replied:
Any downside to using the package module?

mkenigs replied:
Oh yeah that's true. Since analyze is the first step and calculates store paths, what about using store  paths for every step afterwards? That would be (marginally) faster anyways, right?

ysndr replied:
Yeah i just made them all use the installable, rather prefer stop changing things for changing them, if you agree i'd go with how it is now, and we can look at store paths when we have flox-types integrated properly, which would make accessing them safer as well

mkenigs commented on crates/flox-rust-sdk/src/models/publish.rs:
```diff
@@ -226,22 +231,29 @@ impl<'flox> Publish<'flox, NixAnalysis> {
+    pub async fn build(&self) -> Result<(), PublishError> {
+        let nix = self.flox.nix(Default::default());
+        let command = Build {
+            installables: [self.installable()].into(),
```
I think current publish builds all outputs:
https://github.com/flox/flox/blob/eff9f4f0a726cb5ffa632f2672f0577d75becb78/flox-bash/lib/commands/publish.sh#L470

Are we deliberately changing that behavior? Can you add a comment with some explanation of what we've decided?

ysndr replied:
Yeah we should just use `^*` everywhere (also rather than checking individual store paths i think)

ysndr replied:
time to add that to runix

ysndr replied:
flox/runix#47

mkenigs commented on tests/publish.bats:
```diff
@@ -0,0 +1,100 @@
+# The setup will create a channel repo with a `hello` package.
+# `flox publish2` resolves local repositories to their upstream counterpart.
+# Since we don't want to manage and pollute upstream repositories,
+# we set the reposotry as its own remote, and can verify publishes by checking
```
```suggestion
# we set the repository as its own remote, and can verify publishes by checking
```

mkenigs commented on crates/flox-rust-sdk/src/models/publish.rs:
```diff
@@ -290,6 +311,27 @@ impl<'flox> Publish<'flox, NixAnalysis> {
+        &mut self,
+        substituter: SubstituterUrl,
+    ) -> Result<(), PublishError> {
+        let cache_meta = self.get_binary_cache_metadata(Some(substituter)).await?;
+        let cache_meta_json =
+            serde_json::to_value(cache_meta).map_err(PublishError::SerializeCacheMeta)?;
+        let mut caches = self
+            .analysis
+            .0
+            .get("cache")
+            .map(|v| v.as_array().expect("invalid metadata").to_owned())
+            .unwrap_or_default();
+
+        caches.push(cache_meta_json);
```
What was blocking us on using catalog types? This whole method could essentially be dropped

ysndr replied:
I think mainly the custom flakeref representation we have there that is kind of everywhere in the tests and such.
I'd prefer to use the catalog entry type otherwise yes

mkenigs replied:
https://github.com/flox/flox/issues/203

mkenigs commented on crates/flox-rust-sdk/src/models/publish.rs:
```diff
@@ -426,34 +474,47 @@ struct UpstreamCatalog<'a>(&'a Git);
+            .expect("invalid metadata");
+
+        let nix_out_hash = &Path::new(
+            snapshot["element"]["storePaths"][0]
```
Do we preserve order everywhere? Seems a bit unreliable to assume we will

ysndr replied:
We shouldn't be modifying these files anyway, nix depends on the order of these i think and I dont know where to get a "default" output hash from otherwise

mkenigs replied:
Does nix even have `storePaths`? I think it might rely on the order of `outputs`? But I'm a bit skeptical that we just got lucky and happened to both preserve the order of outputs and storePaths. You could just hash the list of all store paths - that would have the additional benefit(?) of preventing collisions between `flox publish attr^out` and `flox publish attr^dev`

ysndr commented on crates/flox/src/commands/package.rs:
```diff
@@ -205,6 +208,34 @@ pub(crate) mod interface {
+        #[bpaf(long, short('s'))]
+        pub substituter_url: Option<SubstituterUrl>,
+
+        /// Print snapshot to stdout instead of oploading it to the catalog
+        ///
+        /// When omitted, uses the value for cache-url and falls back to the config
```
well, you write to a catalog, which is associated with a channel.
Using "channel" somewhat stretches our definition established with `(un)subscribe`.

ysndr commented on crates/flox/src/commands/package.rs:
```diff
@@ -350,22 +381,100 @@ impl PackageCommands {
-                    "{}",
-                    serde_json::to_string_pretty(publish.analyze().await?.analysis())?
+                let original_flakeref = &installable.flakeref;
+                let publish_flakeref = PublishFlakeRef::from_flake_ref(
+                    installable.flakeref.clone(),
+                    &flox,
+                    args.inner.prefer_https,
+                )
+                .await?;
+
+                if &publish_flakeref != original_flakeref {
+                    info!("Resolved {} to {}", original_flakeref, publish_flakeref);
+                }
+
```
I think it could be, the main reason i left it separate was that i initially had `allow_dirty` and `prefer_upstream` flags, which would have pushed `Pulish::new()` to take 7 arguments, 3 of which are not relevant to publishing itself really.
Keeping these steps distinct simplifies the arguments and allows easier error handling IMO

ysndr commented on crates/flox/src/commands/package.rs:
```diff
@@ -350,22 +381,100 @@ impl PackageCommands {
+                        .or(config.flox.cache_url)
+                        .ok_or_else(|| {
+                            anyhow!(indoc! {"
+                    Cache url is required!
+                    Provide using `--cache-url` or the `cache_url` config key
```
Yes, sure that argument part should come first

ysndr commented on crates/flox/src/commands/package.rs:
```diff
@@ -205,6 +208,34 @@ pub(crate) mod interface {
+
+        /// Signing key file to sign the binary with
+        ///
+        /// When omitted, reads from the config
```
The config is kind of its own subsystem with support for env variables, imperative config commands and hierarchical files. Would you want to list all available options here (and likewise any other place where we default from the config)?
I agree we should make the config subsystem more accessible, and I'm wondering how we could properly word that.

For starters we could at least provide a command here e.g.

```suggestion
        /// When omitted, reads from the config
        /// set using `flox config --set sign_key url`
        /// see flox-config(1) for more details.
```

mkenigs replied:
Yep I think that's great - or if that feels to long, I think the important part to add is `see flox-config(1)`

garbas reviewed (commented):
I'm only going to comment on the names of the command line options. As I was just documenting `install-flox-action` I was having strong urge to change names there as well (those referring to caching) since it will make no sense to users who don't read documentation and that's most of them.

I like the option name `--cache-url` since it just makes sense without having to understand any other concept.

`--substituter-url` is a very Nix specific word (`substituter`) and is a word we should avoid at all cost. I'd consult @ghudgins on this one since we need to use consistent naming across our tools. Similar problem with naming caches we also have in `install-flox-action` which I just documented and I don't expect many to have easy time understanding it. Maybe `--cache-download-url` might be a better fit.

I would change `--sign-key` to `--cache-signing-key` indicating that this is the key that signs the cache. We might also consider calling it `--cache-private-key` since there is a public and private part to the key.

`--prefer-https` option is least important and I'd put it last on the list on the `--help` screen.

ysndr commented:
Moved the `--prefer-https` flag, it was the first one i added and new ones just went below, my bad.

I agree that `--substituter-url` is likely not the clearest..

Not sure what is the shortest way to say "alternative url to a cache if the cache is accessed differently by consumers..."

`--cache-download-url` might be a better fit, `--public-cache-url` might work too?

mkenigs commented on crates/flox/src/commands/package.rs:
```diff
@@ -344,28 +341,108 @@ impl PackageCommands {
+                    .with_context(|| format!("Could not sign binary with sign-key {sign_key:?}"))?;
+                info!("done!");
+
+                // caching binary
```
```suggestion
                // cache binary
```

mkenigs commented on tests/publish.bats:
```diff
@@ -0,0 +1,100 @@
+#       snapshots, but it is a minimally invasive solution for testing.
+setup() {
+    # setup up temporary local channel
+    export CHANNEL="$(mktemp -d)"
```
Should you use `BATS_TEST_TMPDIR`? (I'm not honestly sure)

ysndr replied:
Not sure if that is created per test or for the entire suite.

mkenigs commented on tests/publish.bats:
```diff
@@ -0,0 +1,100 @@
+# * Get version and output hash for the `hello` package
+#   to check the generated catalog against.
+setup_file() {
+    nix-serve -p 8081 &
```
Well this is nice and thorough :)

tomberek commented on crates/flox/doc/flox-publish.md:
```diff
@@ -10,12 +10,30 @@ flox-publish - build and publish project to flox channel
+
+The package must be defined in a remote git repository and be referred to
+either directly by a `git+ssh://<url>[#<package>]` url or another url that
+can be resolved to an upstream git resource.
```
needs some note about the heuristic that the remote origin is used

ysndr replied:
```suggestion
can be resolved to an upstream git resource.
Packages in local repositories are resolved to the current branch's upstream branch.
```

tomberek commented on crates/flox/src/commands/package.rs:
```diff
@@ -172,44 +175,42 @@ pub(crate) mod interface {
+        #[bpaf(long, short('s'))]
+        pub public_cache_url: Option<SubstituterUrl>,
+
+        /// Print snapshot JSON to stdout instead of uploading it to the catalog
```
Awkward to have `--json` change behavior in addition to changing the format

tomberek commented on crates/flox/src/commands/package.rs:
```diff
@@ -344,28 +341,108 @@ impl PackageCommands {
+                    .context("Failed uploading binary")?;
+                info!("done!");
+
+                info!("Checking binary can be downloaded from {substituter_url}...");
```
This can easily fail, especially with S3 and when there is a S3/HTTPS split.

tomberek commented on crates/flox/src/commands/package.rs:
```diff
@@ -344,28 +341,108 @@ impl PackageCommands {
+                info!("done!");
+
+                if args.inner.json {
+                    let analysis = publish.analysis();
```
TODO: replace with scrape, or related

tomberek reviewed (approved):
Please leave as `publish2` or a feature-flag/var to ease the impacts on infra. Coordination time!

mkenigs commented:
> Please leave as `publish2` or a feature-flag/var to ease the impacts on infra. Coordination time!

We'll need changes in update-catalog and nixpkgs-flox, floxpkgs/modules, and deltaops
